File: services/auth-service/internal/services/auth_service.go
package services

import (
	"context"
	"fmt"

	"github.com/ronakmalkan/puzzle/auth-service/internal/models"
	"github.com/ronakmalkan/puzzle/auth-service/pkg/jwt"
	"github.com/ronakmalkan/puzzle/auth-service/pkg/password"
)

type SignupInput struct {
	Email     string `json:"email" binding:"required,email"`
	Password  string `json:"password" binding:"required"`
	FirstName string `json:"firstname" binding:"required"`
	LastName  string `json:"lastname" binding:"required"`
}

type LoginInput struct {
	Email    string `json:"email" binding:"required,email"`
	Password string `json:"password" binding:"required"`
}

type AuthResponse struct {
	Token string       `json:"token"`
	User  *models.User `json:"user"`
}

type UserInfo struct {
	ID        int64  `json:"id"`
	Email     string `json:"email"`
	FirstName string `json:"firstname"`
	LastName  string `json:"lastname"`
}

type AuthService struct {
	userRepo   *models.UserRepository
	jwtManager *jwt.JWTManager
}

func NewAuthService(userRepo *models.UserRepository, jwtManager *jwt.JWTManager) *AuthService {
	return &AuthService{
		userRepo:   userRepo,
		jwtManager: jwtManager,
	}
}

func (s *AuthService) Signup(ctx context.Context, input SignupInput) (*AuthResponse, error) {
	// Validate password strength
	if !password.ValidatePasswordStrength(input.Password) {
		return nil, fmt.Errorf("password does not meet security requirements: must be at least 8 characters with uppercase, lowercase, digit, and special character")
	}

	// Check if email already exists
	exists, err := s.userRepo.EmailExists(ctx, input.Email)
	if err != nil {
		return nil, fmt.Errorf("failed to check email existence: %w", err)
	}
	if exists {
		return nil, fmt.Errorf("email already registered")
	}

	// Hash password
	hashedPassword, err := password.HashPassword(input.Password)
	if err != nil {
		return nil, fmt.Errorf("failed to hash password: %w", err)
	}

	// Create user
	user, err := s.userRepo.Create(ctx, models.CreateUserInput{
		Email:     input.Email,
		Password:  hashedPassword,
		FirstName: input.FirstName,
		LastName:  input.LastName,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to create user: %w", err)
	}

	// Generate JWT token
	token, err := s.jwtManager.GenerateToken(jwt.TokenPayload{
		ID:    user.ID,
		Email: user.Email,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to generate token: %w", err)
	}

	return &AuthResponse{
		Token: token,
		User:  user,
	}, nil
}

func (s *AuthService) Login(ctx context.Context, input LoginInput) (*AuthResponse, error) {
	// Find user by email
	user, err := s.userRepo.FindByEmail(ctx, input.Email)
	if err != nil {
		return nil, fmt.Errorf("invalid email or password")
	}

	// Compare password
	match, err := password.ComparePassword(input.Password, user.Password)
	if err != nil {
		return nil, fmt.Errorf("failed to verify password: %w", err)
	}
	if !match {
		return nil, fmt.Errorf("invalid email or password")
	}

	// Generate JWT token
	token, err := s.jwtManager.GenerateToken(jwt.TokenPayload{
		ID:    user.ID,
		Email: user.Email,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to generate token: %w", err)
	}

	return &AuthResponse{
		Token: token,
		User:  user,
	}, nil
}

func (s *AuthService) GetUserInfo(ctx context.Context, userID int64) (*UserInfo, error) {
	user, err := s.userRepo.FindByID(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("failed to get user info: %w", err)
	}

	return &UserInfo{
		ID:        user.ID,
		Email:     user.Email,
		FirstName: user.FirstName,
		LastName:  user.LastName,
	}, nil
}

File: services/auth-service/src/services/auth.service.ts
import * as UserModel from '../models/user.model';
import { hashPassword, comparePassword, validatePasswordStrength } from '../utils/password';
import { validateEmail } from '../utils/validation';
import { generateToken } from '../utils/jwt';
import { logger } from '../utils/logger';

export interface SignupRequest {
  email: string;
  password: string;
  firstname: string;
  lastname: string;
}

export interface LoginRequest {
  email: string;
  password: string;
}

export interface AuthResponse {
  message: string;
  token: string;
  user: {
    id: number;
    email: string;
    firstname: string;
    lastname: string;
  };
}

export const signup = async (data: SignupRequest): Promise<AuthResponse> => {
  const { email, password, firstname, lastname } = data;

  // Validate email format
  if (!validateEmail(email)) {
    throw new Error('Invalid email format. Please enter a valid email address.');
  }

  // Validate password strength
  if (!validatePasswordStrength(password)) {
    throw new Error(
      'Password must be at least 8 characters long and contain at least 1 uppercase letter, 1 lowercase letter, 1 number, and 1 special character.'
    );
  }

  // Check if email already exists
  const emailAlreadyExists = await UserModel.emailExists(email);
  if (emailAlreadyExists) {
    throw new Error('Email already exists. Please use a different email address.');
  }

  // Hash password
  const hashedPassword = await hashPassword(password);

  // Create user
  const user = await UserModel.createUser({
    email,
    password: hashedPassword,
    firstname,
    lastname,
  });

  logger.info({ userId: user.id, email: user.email }, 'User created successfully');

  // Generate JWT token
  const token = generateToken({ id: user.id, email: user.email });

  return {
    message: 'User registered successfully',
    token,
    user: {
      id: user.id,
      email: user.email,
      firstname: user.firstname,
      lastname: user.lastname,
    },
  };
};

export const login = async (data: LoginRequest): Promise<AuthResponse> => {
  const { email, password } = data;

  // Get user by email
  const user = await UserModel.getUserByEmail(email);
  if (!user) {
    throw new Error('Invalid email or password.');
  }

  // Verify password
  const isPasswordValid = await comparePassword(password, user.password);
  if (!isPasswordValid) {
    throw new Error('Invalid email or password.');
  }

  logger.info({ userId: user.id, email: user.email }, 'User logged in successfully');

  // Generate JWT token
  const token = generateToken({ id: user.id, email: user.email });

  return {
    message: 'Login successful',
    token,
    user: {
      id: user.id,
      email: user.email,
      firstname: user.firstname,
      lastname: user.lastname,
    },
  };
};

export const getUserInfo = async (userId: number) => {
  const user = await UserModel.getUserById(userId);

  if (!user) {
    throw new Error('User not found');
  }

  return {
    id: user.id,
    email: user.email,
    firstname: user.firstname,
    lastname: user.lastname,
    created_at: user.created_at,
  };
};
